lib/reporting/sp_proofing_events_by_uuid.rb

```diff
@@ -0,0 +1,205 @@
+                by properties.user_id as login_uuid
+        | filter workflow_started > 0 or verified_by_mail > 0 or verified_fraud_review > 0
+        | limit 10000
+      QUERY
+    end
+
+    def as_csv
+      csv = []
+      csv << ['Date Range', "#{time_range.begin.to_date} - #{time_range.end.to_date}"]
+      csv << csv_header
+      data.each do |result_row|
+        csv << result_row
+      end
+      csv.compact
+    end
+
+    def to_csv
+      CSV.generate do |csv|
+        as_csv.each do |row|
+          csv << row
+        end
+      end
+    end
+
+    def as_emailable_reports
+      [
+        EmailableReport.new(
+          title: "#{agency_abbreviation} Proofing Events By UUID",
+          table: as_csv,
+          filename: "#{agency_abbreviation.downcase}_proofing_events_by_uuid",
+        ),
+      ]
+    end
+
+    def csv_header
+      [
+        'UUID',
+        'Workflow Started',
+        'Documnet Capture Started',
+        'Document Captured',
+        'Selfie Captured',
+        'Document Authentication Passed',
+        'SSN Submitted',
+        'Personal Information Submitted',
+        'Personal Information Verified',
+        'Phone Submitted',
+        'Phone Verified',
+        'Verification Workflow Complete',
+        'Identity Verified for In-Band Users',
+        'Identity Verified for Verify-By-Mail Users',
+        'Identity Verified for Fraud Review Users',
+        'Out-of-Band Verification Pending Seconds',
+        'Agency Handoff Visited',
+        'Agency Handoff Submitted',
+      ]
+    end
+
+    def data
+      @data ||= fetch_results.map do |result_row|
+        process_result_row(result_row)
+      end
+    end
+
+    def process_result_row(result_row)
+      login_uuid = result_row['login_uuid']
+      agency_uuid = convert_uuid(login_uuid)
+      return unless agency_uuid.present?
+      [
+        agency_uuid,
+        result_row['workflow_started'] == '1',
+        result_row['doc_auth_started'] == '1',
+        result_row['document_captured'] == '1',
+        result_row['selfie_captured'] == '1',
+        result_row['doc_auth_passed'] == '1',
+        result_row['ssn_submitted'] == '1',
+        result_row['personal_info_submitted'] == '1',
+        result_row['personal_info_verified'] == '1',
+        result_row['phone_submitted'] == '1',
+        result_row['phone_verified'] == '1',
+        result_row['online_workflow_completed'] == '1',
+        result_row['verified_in_band'] == '1',
+        result_row['verified_by_mail'] == '1',
+        result_row['verified_fraud_review'] == '1',
+        result_row['out_of_band_verification_pending_seconds'].to_i,
+        result_row['agency_handoff'] == '1',
+        result_row['sp_redirect'] == '1',
+      ]
+    end
+
+    def convert_uuid(uuid)
+      user = User.find_by(uuid: uuid)
+      user&.agency_identities&.find_by(agency:)&.uuid
+    end
```

Comment:
What's the rough size of these reports? My first instinct would be to batch these and memoize them as a map this is like 3 queries per UUID

Something along the lines of...

```ruby
  def build_uuid_map(uuids)
    uuid_map = Hash.new

    AgencyIdentity
      .joins(:user)
      .where(
        agency:,
        users: { uuid: uuids }
      ).in_batches do |batch|
        batch.each do |agency_identity|
          uuid_map[agency_identity.user.uuid] = agency_identity.uuid
        end
      end

    uuid_map
  end
```

I couldn't figure out an easy way to pare down the `User` object to user `id/uuid` (like normally `User.select(:id, :uuid)` to keep it lighter in memory, but this should do an entire batch as one query

Reply:
They usually have 4k-6k users. I had the thought to do a smaller number of big queries as well. I dropped it to try and get the report working; I can take another look at it now that I (hopefully) do have the report working.